### PR TITLE
Issue 926: transition domain object creation

### LINKED ISSUE #926: Data Migration: transition_domain domain object ingest and create/update
### Issue Description

Create a script that takes all the data from the transition_domain table from #923 and makes the corresponding objects in our main tables.
The script should:
- Creates a domain object that does NOT have a corresponding application for each domain listed in the table
  - it should not create duplicates
  - if a domain already exists, it should check the domain.state and if its not equal to transition_domain.status, set it to be the new value
- Creates a domainInvitation object for each of the emails in the username field and associates it with the corresponding domain object
  - emails should **NOT** be sent upon creation of the domain invitation objects
- This must be able to run in combination with the docker-compose run app ./manage.py "script_name"

### Additional Context (optional)

https://docs.google.com/document/d/1UkQZdAmcGprXWdHX_PuT155CCNPbf_Gf/edit?pli=1#heading=h.iqv2v8lecjpn

### Issue Links

Depends on: #923

### PR DESCRIPTION
**Pull request by CocoByte**
Resolves 
#926 

## Changes

- Created script transfer_transition_domains_to_domains.py
- Updated model domain.py (to create workaround for executing updates to States without executing EPP calls)
- Updated model transition_domain.py (to fix a bug where "Hold" should be "On Hold")
- Created migration for model updates

## Setup

    
   - [ ] Run `docker-compose up` and navigate to /admin
   - [ ] Create some transition domains (see example below)
Select transition domain to change

### For issue #923 
  Open a terminal and run the following commands:
   
`cd src` (navigate to the src folder)

`docker compose run -T app ./manage.py transfer_transition_domains_to_domains --debug`

**NOTE**: 
`--debug` will print out additional, detailed logs

#### 1. Verify that the script...
- [ ] Adds new Domain entries (Check the Domain table on the admin page)
- [ ] Adds new Domain Invitation entries (check the Domain Invitation table on the admin page)
Pasted Graphic 1

#### 2. Re-execute the docker command and verify that the script...
- [ ] Does not add or update any data
Pasted Graphic 2

#### 3. Change some values in the test data (via the admin page) and verify that the script...
- [ ] Updates the corresponding entries

_In the example below, changed "bestCampsites" website from **OnHold** to **Ready**.
Re-ran the script an the output shows 1 updated domain_

Select transition domain to change

Pasted Graphic 5

#### Satisfied acceptance criteria and met development standards

- [x] Met the acceptance criteria, or will meet them in a subsequent PR
- [ ] Created/modified automated tests
- [x] Added at least 2 developers as PR reviewers (only 1 will need to approve)
- [x] Messaged on Slack or in standup to notify the team that a PR is ready for review
- [ ] Changes to “how we do things” are documented in READMEs and or onboarding guide
- [x] If any model was updated to modify/add/delete columns, makemigrations was ran and the assoicated migrations file has been commited.

#### Ensured code standards are met (Original Developer)

- [x] All new functions and methods are commented using plain language
- [ ] Did dependency updates in Pipfile also get changed in requirements.txt?
- [ ] Interactions with external systems are wrapped in try/except
- [x] Error handling exists for unusual or missing values

#### Reviewed, tested, and left feedback about the changes

- [ ] Pulled this branch locally and tested it
- [ ] Reviewed this code and left comments
- [ ] Checked that all code is adequately covered by tests
- [ ] Made it clear which comments need to be addressed before this work is merged
- [ ] If any model was updated to modify/add/delete columns, makemigrations was ran and the assoicated migrations file has been commited.

#### Ensured code standards are met (Code reviewer)

- [ ] All new functions and methods are commented using plain language
- [ ] Interactions with external systems are wrapped in try/except
- [ ] Error handling exists for unusual or missing values
- [ ] (Rarely needed) Did dependency updates in Pipfile also get changed in requirements.txt?

#### Validated user-facing changes as a developer

- [ ] New pages have been added to .pa11yci file so that they will be tested with our automated accessibility testing
- [ ] Checked keyboard navigability
- [ ] Meets all designs and user flow
- [ ] Tested general usability, landmarks, page header structure, and links with a screen reader (such as Voiceover or ANDI)

- [ ] Tested with multiple browsers, the suggestion is to use ones that the developer didn't (check off which ones were used)
  - [ ] Chrome
  - [ ] Microsoft Edge
  - [ ] FireFox
  - [ ] Safari

- [ ] (Rarely needed) Tested as both an analyst and applicant user

**Note:** Multiple code reviewers can share the checklists above, a second reviewers should not make a duplicate checklist

#### Verified that the changes match the design intention

- [ ] Checked that the design translated visually
- [ ] Checked behavior
- [ ] Checked different states (empty, one, some, error)
- [ ] Checked for landmarks, page heading structure, and links
- [ ] Tried to break the intended flow

#### Validated user-facing changes as a designer

- [ ] Checked keyboard navigability
- [ ] Tested general usability, landmarks, page header structure, and links with a screen reader (such as Voiceover or ANDI)

- [ ] Tested with multiple browsers (check off which ones were used)
  - [ ] Chrome
  - [ ] Microsoft Edge
  - [ ] FireFox
  - [ ] Safari

- [ ] (Rarely needed) Tested as both an analyst and applicant user

### CODE REVIEW

File: src/registrar/models/transition_domain.py [hunk 1]
```diff
@@ -5,14 +5,18 @@
 
 class StatusChoices(models.TextChoices):
     READY = "ready", "Ready"
```
Comment (abroddrick) [inline src/registrar/models/transition_domain.py:7]: Just realized this will make your other PR fail, this will definitely have to be kept in mind

File: src/registrar/management/commands/transfer_transition_domains_to_domains.py [hunk 2]
```diff
@@ -0,0 +1,229 @@
...
+    def add_arguments(self, parser):
+        parser.add_argument("--debug", action=argparse.BooleanOptionalAction)
+
+        parser.add_argument(
+            "--limitParse", default=0, help="Sets max number of entries to load"
+        )
+
+    def handle(  # noqa: C901
+        self,
+        **options,
```
Comment (abroddrick) [inline src/registrar/management/commands/transfer_transition_domains_to_domains.py:48]: like with the other PR I would ask that this get broken up into other functions to make it a bit more readable.

File: src/registrar/management/commands/transfer_transition_domains_to_domains.py [hunk 3]
```diff
@@ -0,0 +1,229 @@
...
+        Domain.objects.bulk_create(to_create)
+
+        total_new_entries = len(to_create)
+        total_updated_domain_entries = len(updated_domain_entries)
+
+        logger.info(
+            f"""{termColors.OKGREEN}
+
+============= FINISHED ===============
+Created {total_new_entries} transition domain entries,
```
Comment (abroddrick) [inline src/registrar/management/commands/transfer_transition_domains_to_domains.py:178]: can we fix this indenting? and the other ones below that do the same thing

File: src/registrar/management/commands/transfer_transition_domains_to_domains.py [hunk 4]
```diff
@@ -0,0 +1,229 @@
+"""Load domain invitations for existing domains and their contacts."""
```
Comment (abroddrick) [inline src/registrar/management/commands/transfer_transition_domains_to_domains.py:1]: remove comment

**Review by abroddrick — COMMENTED** [review]
Didn't run or thoroughly review, just leaving comments for now

**Comment by CocoByte** [comment]
I will merge from Main to fix failed tests.  But ALSO need to add domainInvitation piece, which I missed.  Will post in slack when this is updated.

**Review by zandercymatics — COMMENTED** [review]
One thing to note is that while testing step 3,  if you change the domain name on an existing domain (for instance, test.gov => test123.gov), it will show up in the created column rather than the updated one. It works for every other field, though I am not sure if this is a bug or intended behavior. If its intended I'll approve
image

File: src/registrar/models/transition_domain.py (hunk 1 above)
Comment (CocoByte) [reply src/registrar/models/transition_domain.py:7]: Good catch. In my latest updates I matched the transitionDomain model in both branches (along with the migration) so hopefully there will be no collisions there.

File: src/registrar/management/commands/transfer_transition_domains_to_domains.py (hunk 3 above)
Comment (CocoByte) [reply src/registrar/management/commands/transfer_transition_domains_to_domains.py]: Indents fixed

File: src/registrar/management/commands/transfer_transition_domains_to_domains.py (hunk 4 above)
Comment (CocoByte) [reply src/registrar/management/commands/transfer_transition_domains_to_domains.py:1]: Done

File: src/registrar/management/commands/transfer_transition_domains_to_domains.py [hunk 5]
```diff
@@ -0,0 +1,229 @@
...
+                if debug_on:
+                    logger.info(
+                        f"""{termColors.WARNING}
+    > Found existing domain entry for: {transition_domain_name}, {current_state}
+    {termColors.ENDC}"""
+                    )
+                if transition_domain_status != current_state:
+                    if (
+                        transition_domain_status
+                        == TransitionDomain.StatusChoices.ON_HOLD
```
[the remark this answers is not in the document]
Comment (CocoByte) [reply src/registrar/management/commands/transfer_transition_domains_to_domains.py:140]: yes

File: src/registrar/management/commands/transfer_transition_domains_to_domains.py [hunk 6]
```diff
@@ -0,0 +1,229 @@
...
+
+        total_new_entries = len(to_create)
+        total_updated_domain_entries = len(updated_domain_entries)
+
+        logger.info(
+            f"""{termColors.OKGREEN}
+
+============= FINISHED ===============
+Created {total_new_entries} transition domain entries,
+updated {total_updated_domain_entries} transition domain entries
```
[the remark this answers is not in the document]
Comment (CocoByte) [reply src/registrar/management/commands/transfer_transition_domains_to_domains.py]: I like it.  Implementing

File: src/registrar/management/commands/transfer_transition_domains_to_domains.py (hunk 2 above)
Comment (CocoByte) [reply src/registrar/management/commands/transfer_transition_domains_to_domains.py:173]: Done

File: src/registrar/management/commands/transfer_transition_domains_to_domains.py [hunk 7]
```diff
@@ -0,0 +1,323 @@
...
+                # DEBUG:
+                self.print_debug(
+                    debug_on,
+                    f"""{termColors.OKCYAN}
+                    Processing Transition Domain: {transition_domain_name}, {transition_domain_status}, {transition_domain_email}
+                    {termColors.ENDC}""",  # noqa
+                )
+
+                # get the existing domain
+                target_domain = Domain.objects.get(name=transition_domain_name)
```
Comment (abroddrick) [inline src/registrar/management/commands/transfer_transition_domains_to_domains.py:224]: Why are we not using .exists() here?

File: src/registrar/management/commands/transfer_transition_domains_to_domains.py [hunk 8]
```diff
@@ -0,0 +1,323 @@
...
+                # for existing entry, update the status to
+                # the transition domain status
+                update_made = self.update_domain_status(
+                    transition_domain, target_domain, debug_on
+                )
+                if update_made:
+                    updated_domain_entries.append(transition_domain.domain_name)
+
+            except Domain.DoesNotExist:
+                already_in_to_create = next(
```
Comment (abroddrick) [inline src/registrar/management/commands/transfer_transition_domains_to_domains.py:243]: This is a bit hard to read I would add a comment here

File: src/registrar/management/commands/transfer_transition_domains_to_domains.py [hunk 9]
```diff
@@ -0,0 +1,323 @@
...
+                        {termColors.ENDC}""",
+                    )
+                else:
+                    # no matching entry, make one
+                    new_entry = Domain(
+                        name=transition_domain_name, state=transition_domain_status
+                    )
+                    domains_to_create.append(new_entry)
+
+                    if transition_domain_email:
```
Comment (abroddrick) [inline src/registrar/management/commands/transfer_transition_domains_to_domains.py:263]: x is not a boolean you want to avoid checking Truthy in python as a rule. I would check against None/empty string vs doing the truthy check

File: src/registrar/management/commands/transfer_transition_domains_to_domains.py [hunk 10]
```diff
@@ -0,0 +1,323 @@
...
+                    > Found existing domain entry for: {transition_domain_name}, {target_domain.state}
+                    {termColors.ENDC}""",  # noqa
+                )
+
+                # for existing entry, update the status to
+                # the transition domain status
+                update_made = self.update_domain_status(
+                    transition_domain, target_domain, debug_on
+                )
+                if update_made:
```
Comment (abroddrick) [inline src/registrar/management/commands/transfer_transition_domains_to_domains.py:239]: We can have duplicate domains with different users. We should be making a domain invitation for each user, even if the domain object exists. Hence since we have to repeat some of the code that is below, it may be better to do .exists() instead of waiting for an exception.

File: src/registrar/management/commands/transfer_transition_domains_to_domains.py [hunk 11]
```diff
@@ -0,0 +1,323 @@
...
+                        new_domain_invitation = DomainInvitation(
+                            email=transition_domain_email.lower(), domain=new_entry
+                        )
+                        domain_invitations_to_create.append(new_domain_invitation)
+                    else:
+                        logger.info(
+                            f"{termColors.FAIL} ! No e-mail found for domain: {new_entry}"  # noqa
+                            f"(SKIPPED ADDING DOMAIN INVITATION){termColors.ENDC}"
+                        )
+                        skipped_domain_invitations.append(transition_domain_name)
```
Comment (abroddrick) [inline src/registrar/management/commands/transfer_transition_domains_to_domains.py:273]: since transition domains can have duplicates, couldnt a domain have one row with an actual email and another row with a missing email. In which case we don't want to add to skipped domain invitations if at least SOMEONE got an invitation for this domain. Would require 1.) when adding a domain invitation if that domain exists in skipped_domain_invitations make  remove it from skipped. 2.) Visa versa, don't add to skipped if it is in domain_invitiations_to_create

File: src/registrar/management/commands/transfer_transition_domains_to_domains.py [hunk 12]
```diff
@@ -0,0 +1,323 @@
...
+                debug_max_entries_to_parse > 0
+                and total_rows_parsed >= debug_max_entries_to_parse
+            ):
+                logger.info(
+                    f"""{termColors.YELLOW}
+                    ----PARSE LIMIT REACHED.  HALTING PARSER.----
+                    {termColors.ENDC}
+                    """
+                )
+                break
```
Comment (abroddrick) [inline src/registrar/management/commands/transfer_transition_domains_to_domains.py:311]: dont think this is needed

File: src/registrar/management/commands/transfer_transition_domains_to_domains.py [hunk 13]
```diff
@@ -0,0 +1,323 @@
+"""Load domain invitations for existing domains and their contacts."""
```
Comment (abroddrick) [inline src/registrar/management/commands/transfer_transition_domains_to_domains.py:1]: not existing domains, maybe reword "Load domain Invitations and Domain objects from the transition domain table"

**Review by abroddrick — CHANGES_REQUESTED** [review]
Hate to do it, but have to request changes. There is a big where if a domain has multiple users we aren't making domain invitations for all those users. See example below, bob should have 3 invitations.
image

besides that just readme, and only minor suggestion changes that are non blocking.

File: src/registrar/management/commands/transfer_transition_domains_to_domains.py (hunk 7 above)
Comment (CocoByte) [reply src/registrar/management/commands/transfer_transition_domains_to_domains.py:224]: That would be the proper way to do this.  I think part of me thought it was less efficient, checking if an object exists and then grabbing it is traversing the table twice...and makes this function more complex...plus we check for other exceptions anyways.  But I agree this isn't proper form so I will change it.

File: src/registrar/management/commands/transfer_transition_domains_to_domains.py (hunk 8 above)
Comment (CocoByte) [reply src/registrar/management/commands/transfer_transition_domains_to_domains.py:243]: I'm going to change this to use .exists() in an if-else statement

File: src/registrar/management/commands/transfer_transition_domains_to_domains.py (hunk 9 above)
Comment (CocoByte) [reply src/registrar/management/commands/transfer_transition_domains_to_domains.py:263]: good to know.  I'll update

File: src/registrar/management/commands/transfer_transition_domains_to_domains.py (hunk 10 above)
Comment (CocoByte) [reply src/registrar/management/commands/transfer_transition_domains_to_domains.py:239]: K - will update the algorithm

File: src/registrar/management/commands/transfer_transition_domains_to_domains.py (hunk 12 above)
Comment (CocoByte) [reply src/registrar/management/commands/transfer_transition_domains_to_domains.py:311]: needed to exit loop if parse limit has been reached

File: src/registrar/management/commands/transfer_transition_domains_to_domains.py (hunk 13 above)
Comment (CocoByte) [reply src/registrar/management/commands/transfer_transition_domains_to_domains.py:1]: I thought I removed this comment...must have re-emerged in the course of my commits

**Comment by abroddrick** [comment]
> One thing to note is that while testing step 3, if you change the domain name on an existing domain (for instance, test.gov => test123.gov), it will show up in the created column rather than the updated one. 

I think that should happen, because the domain name is really our unique id we get from the files.

**Review by abroddrick — APPROVED** [review]
Please at least react or reply to existing comments that have not received a response. But none are blocking and I see now invitations are made for all users, this works for me
